iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -0,0 +1,309 @@
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowTypes.h"
+#include "iree/compiler/Dialect/Flow/Transforms/DestructiveUpdateUtils.h"
+#include "iree/compiler/Dialect/Shape/IR/ShapeDialect.h"
+#include "iree/compiler/Dialect/Shape/IR/ShapeOps.h"
+#include "llvm/ADT/STLExtras.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/Block.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/TypeRange.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-linalg-tile-and-distribute-on-tensors"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+struct DispatchLinalgOnTensorsPass
+    : public PassWrapper<DispatchLinalgOnTensorsPass, OperationPass<FuncOp>> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, IREE::Flow::FlowDialect,
+                    AffineDialect, scf::SCFDialect, ShapeDialect>();
+  }
+  DispatchLinalgOnTensorsPass() = default;
+  DispatchLinalgOnTensorsPass(ArrayRef<int64_t> sizes) {
+    this->tileSizes = sizes;
+  };
+  DispatchLinalgOnTensorsPass(const DispatchLinalgOnTensorsPass &pass) {}
+  void runOnOperation() override;
+
+ private:
+  ListOption<int64_t> tileSizes{
+      *this, "tile-sizes", llvm::cl::desc("Set tile sizes to use"),
+      llvm::cl::ZeroOrMore, llvm::cl::MiscFlags::CommaSeparated};
+};
+
+static Operation *buildFlowWorkgroupDispatchOp(OpBuilder &b,
+                                               linalg::LinalgOp root,
+                                               linalg::LinalgOp &clonedRoot) {
+  Location loc = root->getLoc();
+  // TODO(nicolasvasilache): for now this is a 3-D grid of 1's.
+  // In the future make constant and bind late into backend-specific values.
+  SmallVector<Value, 1> count(3, b.create<ConstantIndexOp>(loc, 1));
+
+  SmallVector<Value, 4> dispatchOperands;
+  dispatchOperands.reserve(root->getNumOperands());
+  dispatchOperands.append(root.getInputs().begin(), root.getInputs().end());
+  for (OpOperand &opOperand : root.getOutputOpOperands()) {
+    if (root.isInitTensor(&opOperand)) {
```

Comment:
Not really sure why the operand value is used if this is an initTensor (you only need the shape then), but only the shape is needed if it is not. Is the condition reversed here or am I missing something?

Reply:
The core refactoring made `isInitTensor` a property that we check on output tensors.
If the output tensor has all the right properties **and** the corresponding scalar is used in the linalg.generic region, then `isInitTensor` returns true.